main.rb

```diff
@@ -32,51 +45,159 @@ def calc_percent(numerator, denominator)
+      count += 1
+    else
+      puts("Error while checking plan status. Response from server: #{get_parsed_response(res.body)}")
+      count += 1
+    end
   end
-  JSON.parse(res.body, symbolize_names: true)
+  exit(1)
 end
 
 def find_project(access_token)
-  uri = URI.parse("https://testinium.io/Testinium.RestApi/api/projects/#{$project_id}")
-  req = Net::HTTP::Get.new(uri.request_uri,
-                           { 'Content-Type' => 'application/json', 'Authorization' => "Bearer #{access_token}" })
-  res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
-    http.request(req)
+  is_success = false
+  count = 1
+  puts("Starting to find the project...")
+
+  while !is_success && is_count_less_than_max_api_retry(count) do
+    check_timeout()
+    puts("Finding project. Number of attempts: #{count}")
+
+    uri = URI.parse("https://testinium.io/Testinium.RestApi/api/projects/#{$project_id}")
+    req = Net::HTTP::Get.new(uri.request_uri,
+                             { 'Content-Type' => 'application/json', 'Authorization' => "Bearer #{access_token}", 'current-company-id' => "#{$company_id}" })
+    res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
+      http.request(req)
+    end
+
+    if (res.kind_of? Net::HTTPSuccess)
+      puts('Project was found successfully...')
+      return get_parsed_response(res.body)
+    elsif (res.kind_of? Net::HTTPClientError)
+      puts(get_parsed_response(res.body)[:message])
+      count += 1
+    else
+      puts("Error while finding project. Response from server: #{get_parsed_response(res.body)}")
+      count += 1
+    end
+  end
+  exit(1)
+end
+
+def upload(access_token)
+  is_success = false
+  count = 1
+
+  while !is_success && is_count_less_than_max_api_retry(count) do
+    check_timeout()
+    puts("Uploading #{$file_name} to Testinium... Number of attempts: #{count}")
+
+    uri = URI.parse('https://testinium.io/Testinium.RestApi/api/file/upload')
+    req = Net::HTTP::Post.new(uri.request_uri,
+                              { 'Accept' => '*/*', 'Authorization' => "Bearer #{access_token}", 'current-company-id' => "#{$company_id}" })
+    form_data = [
+      ['file', File.open($file)],
+      %w[isSignRequired true]
+    ]
+    req.set_form(form_data, 'multipart/form-data')
+    res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
+      http.request(req)
+    end
+
+    if (res.kind_of? Net::HTTPSuccess)
+      puts('File uploaded successfully...')
+      return get_parsed_response(res.body)
+    elsif (res.kind_of? Net::HTTPClientError)
+      puts(get_parsed_response(res.body)[:message])
+      count += 1
+    else
+      puts("Error while uploading File. Response from server: #{get_parsed_response(res.body)}")
+      count += 1
+    end
   end
-  JSON.parse(res.body, symbolize_names: true)
+  exit(1)
 end
 
 def update_project(project, file_response, access_token)
+  is_success = false
```

Comment:
We do not need the `is_success` variable. "update_project()" function will work fine without it.